### PR TITLE
Hall (New Talk): the PR

### PR DESCRIPTION
**Pull request by Fang-**
Nearly eight months later, it's finally here!

`talk` as we know and love it has been split into `hall` (the backend) and `talk` (the cli frontend). Structures and interfaces got overhauled, library and marks redone, and dependent apps have been updated to match. Fora now has its own super simple app, and uses `hall` to re-publish the content to its own special-purpose circles.

Contains updated JS files for webtalk. PRs for webtalk and fora are on their way.

There's still a lot of things I'd like to change/improve upon, but I'm confident enough in the current state of things that I want to try getting this into the breach. No doubt the chaos of actual usage will reveal some more kinks, but those can be easily patched over the air. I'm looking forward to finally getting this into the hands of the people.

(I noticed just now that some files also have changes to remove trailing whitespace. Didn't intend for that to be included in my commits here, but considering this currently merges cleanly into master, that probably isn't too big a deal.)

Depends on #451.  
Contains (temporary?) workarounds for #447 (investigation ongoing).

Questions and concerns very much welcome.

### CODE REVIEW

**Comment by ohAitch** [comment]
Just to double check,

<details><summary>"doing  =,  talk  on the top-level of this /lib file causes a bail: fail (^:"</summary>

```
~palnup_livpyl= i'm still getting a weird find.$.+2 on a simple json reparser
~palnup_livpyl= like this:
~palnup_livpyl= (of lin+(ot pat+bo txt+so ~) ~)
~palnup_livpyl= it points to (ot ...) specifically, but i see nothing wrong with
-------[61.160]
~palnup_livpyl= it
~pittyp-datfyn# (of lin+(ot pat+bo txt+so ~) ~):jo__<1.lrf {jon/#1 <1.fhl {wer
~palnup_livpyl= yeah it compiles just fine out of context, but there's no part
~palnup_livpyl= of the context i can find that's messing this up
~palnup_livpyl= nevermind, i lied
-------[61.165]
~pittyp-datfyn= huh, are  of ot bo so  all the right things?
~palnup_livpyl= i was doing =,  dejson-soft  =,  talk
~palnup_livpyl= doing that the other way around fixed it
~palnup_livpyl= even though there's nothing in /sur/talk that overwrites any of
~palnup_livpyl= those?
-------[61.170]
~palnup_livpyl= (this is also after putting those in arms because doing =,  talk
~palnup_livpyl= on the top-level of this /lib file causes a bail: fail (^: )
~pittyp-datfyn= uh
~pittyp-datfyn really needs to fix lib/ sur/ handling
~palnup_livpyl= yeah i figured it was some in-dev weirdness
-------[61.175]
~palnup_livpyl= hm, getting the same bail: fail now whenever i try to use a mark
~palnup_livpyl= that calls the lib from earlier
~palnup_livpyl= i guess this means they both compile, hurrah, but pff
~palnup_livpyl= removing the stray =, talk at the top of the mark doesn't
~palnup_livpyl= resolve this either
-------[61.180]
~palnup_livpyl= "port to remaint" they said, "it'll be cool" they said
```

</details>

from the breach checklist to-file section got resolved yes?

**Comment by cgyarvin** [comment]
If you have to ask, just try it :-)

**Comment by Fang-** [comment]
I can put the `=,` on the top level in `/lib/hall-json` just fine now, but the marks still suffer from problems when using `=,`.

For example, changing [this](https://github.com/Fang-/arvo/blob/remaint-merge-talk/mar/hall/prize.hoon#L13) to use `=,` still results in a `bail: fail` as was previously discussed.

**Comment by belisarius222** [comment]
@Fang- there are some conflicts on this PR. I'm about to test it out using the urbit/fleet repo.

**Comment by Fang-** [comment]
Fixed a small regression and tested + improved client behavior on moons. Also removed the lib file's public domain notice, since I expect that to have to go.

As far as I'm concerned, this is once again ready to be merged. @belisarius222 may want to test moons himself to confirm my fixes.

Consider this a request for comments. If anyone's opposed to any part of this making it into the next breach, or feels like something crucial is still missing, please speak up now. (:

File: lib/hall-json.hoon [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+          (cook some ;~(pfix fas pont))
+          (easy ~)
+        ==
+    ;~  pose
+      (cook some ;~(plug ;~(pfix fas pont) -))
+      (easy ~)
+    ==
+  ::
+  ++  sorc                                              ::>  source
+    ;~(plug circ rang)
```
Comment (ohAitch) [inline lib/hall-json.hoon]:
```
++  pont
  ;~  pfix  fas
    %+  sear
      |=  a=coin  ^-  (unit place)
      ?+(a ~ [%$ %da @da] `p.a, [%$ %ud @ud] `p.a)
    nuck:so
  ==
++  sorc  ;~(plug circ (punt ;~(plug pont (punt pont))))
```

File: sur/hall.hoon [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+::      but location in control flow already indicates delta-ness.
+::
+::>  ||
+::>  ||  %wrappers
+::>  ||
+::>    wrapper molds, for semantic clarity.
+::+|
+::
+::TODO  rename
+++  naem  term                                          ::<  circle name
```
[the remark this answers is not in the document]
Comment (Fang-) [reply sur/hall.hoon]: Feels pretty ambiguous with `circle` itself imo. "A `circle` is a `ship` and a `circ`" is weird.

Not even sure if I still want to see this renamed...

File: lib/hall-json.hoon (hunk 1 above)
Comment (Fang-) [reply lib/hall-json.hoon]: That's much more concise, thanks!

File: app/talk.hoon [hunk 3]
```diff
@@ -1,272 +1,654 @@
...
+  [~ ..prep(+<+ u.old)]
+::
+::>  ||
+::>  ||  %utility
+::>  ||
+::>    small utility functions.
+::+|
+::
+++  our
+  (true-self our.bol)
```
[the remark this answers is not in the document]
Comment (Fang-) [reply app/talk.hoon]: Hmm, I guess that's fair. This was intended as "if you're using `our`, you want to use your planet", but I agree using a 3lw may still be weird. `self` has some minor semantic clashing (aka `this`), but is probably fine here.

File: sur/hall.hoon (hunk 2 above)
[the remark this answers is not in the document]
Comment (ohAitch) [reply sur/hall.hoon]: I think the default for not finding a good name should be leaving it as `term`/`@tas`, not introducing phonetic ambiguity. Is just `name` even used anywhere?

[the remark this answers is not in the document]
Comment (Fang-) [reply sur/hall.hoon]: It is, in zuse. When I wrote this, that wasn't namespaced yet. I guess I have no reason not to rename this to `++name` now that it is.

[the remark this answers is not in the document]
Comment (ohAitch) [reply sur/hall.hoon]: ah, the ames one. I'd still push for `chan` or sth that points that this is a chat room name, but either way definitely not `naem`

**Comment by cgyarvin** [comment]
Yeah naem looks like something I’d do as a debugging hack, then clear up later...

**Pull request merged**